modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/security/permission/contributor/SegmentsEntryRoleContributor.java

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.segments.internal.security.permission.contributor;
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.model.User;
+import com.liferay.portal.kernel.security.permission.contributor.RoleCollection;
+import com.liferay.portal.kernel.security.permission.contributor.RoleContributor;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.service.ServiceContextThreadLocal;
+import com.liferay.segments.context.RequestContextMapper;
+import com.liferay.segments.model.SegmentsEntryRole;
+import com.liferay.segments.provider.SegmentsEntryProviderRegistry;
+import com.liferay.segments.service.SegmentsEntryRoleLocalService;
+import com.liferay.segments.simulator.SegmentsEntrySimulator;
+
+import java.util.List;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+
+/**
+ * @author Drew Brokke
+ */
+@Component(immediate = true, service = RoleContributor.class)
+public class SegmentsEntryRoleContributor implements RoleContributor {
+
+	@Override
+	public void contribute(RoleCollection roleCollection) {
+		for (long segmentsEntryId : _getSegmentsEntryIds(roleCollection)) {
+			List<SegmentsEntryRole> segmentsEntryRoles =
+				_segmentsEntryRoleLocalService.getSegmentsEntryRoles(
+					segmentsEntryId);
+
+			for (SegmentsEntryRole segmentsEntryRole : segmentsEntryRoles) {
+				roleCollection.addRoleId(segmentsEntryRole.getRoleId());
+			}
+		}
+	}
+
+	private long[] _getSegmentsEntryIds(RoleCollection roleCollection) {
+		long[] segmentsEntryIds = new long[0];
+
+		ServiceContext serviceContext =
+			ServiceContextThreadLocal.getServiceContext();
+
+		if (serviceContext == null) {
+			return segmentsEntryIds;
+		}
+
+		HttpServletRequest httpServletRequest = serviceContext.getRequest();
```

Comment:
this is weird for me. I'd rather don't use a servletrequest within a service. In our case, we have a similar code but in a PRE-ACTION listener

Reply:
@Reference(
		cardinality = ReferenceCardinality.OPTIONAL,
		policy = ReferencePolicy.DYNAMIC,
		policyOption = ReferencePolicyOption.GREEDY
	)	
	private Supplier<HttpServletRequest> _httpServletRequestSupplier;